File: api/goals.py
import services.goal_service as service
from api.responses import ok, fail


def get_goals(user_id):
    try:
        return ok(service.get_goals(user_id))
    except ValueError as e:
        return fail(e)


def get_goal(goal_id):
    try:
        return ok(service.get_goal(goal_id))
    except ValueError as e:
        return fail(e)


def create_goal(user_id, name, target_amount, target_date=None):
    try:
        return ok(service.create_goal(user_id, name, target_amount, target_date=target_date))
    except ValueError as e:
        return fail(e)


def update_goal(goal_id, name=None, target_amount=None, target_date=None, status=None):
    try:
        return ok(
            service.update_goal(
                goal_id,
                name=name,
                target_amount=target_amount,
                target_date=target_date,
                status=status,
            )
        )
    except ValueError as e:
        return fail(e)


def complete_goal(goal_id):
    try:
        return ok(service.complete_goal(goal_id))
    except ValueError as e:
        return fail(e)


def get_goal_accounts(goal_id):
    try:
        return ok(service.get_goal_accounts(goal_id))
    except ValueError as e:
        return fail(e)


def add_account_to_goal(goal_id, account_id, allocated_amount=0):
    try:
        return ok(service.add_account_to_goal(goal_id, account_id, allocated_amount=allocated_amount))
    except ValueError as e:
        return fail(e)


def update_goal_account_allocation(goal_id, account_id, allocated_amount):
    try:
        return ok(service.update_goal_account_allocation(goal_id, account_id, allocated_amount))
    except ValueError as e:
        return fail(e)


def remove_account_from_goal(goal_id, account_id):
    try:
        return ok(service.remove_account_from_goal(goal_id, account_id))
    except ValueError as e:
        return fail(e)

File: frontend/src/lib/api/goals.ts
import { callApi } from './client';

export type GoalStatus = 'active' | 'completed';

export type Goal = {
	goal_id: number;
	user_id: number;
	name: string;
	target_amount_minor: number;
	target_date?: string | null;
	status: GoalStatus;
};

export type GoalAccount = {
	goal_id: number;
	account_id: number;
	allocated_amount_minor: number;
};

export function getGoals(userId: number) {
	return callApi<Goal[]>('get_goals', userId);
}

export function getGoal(goalId: number) {
	return callApi<Goal | null>('get_goal', goalId);
}

export function createGoal(payload: {
	user_id: number;
	name: string;
	target_amount: number | string;
	target_date?: string | null;
}) {
	return callApi<{ goal_id: number }>(
		'create_goal',
		payload.user_id,
		payload.name,
		payload.target_amount,
		payload.target_date ?? null
	);
}

export function updateGoal(
	goalId: number,
	payload: {
		name?: string | null;
		target_amount?: number | string | null;
		target_date?: string | null;
		status?: GoalStatus | null;
	} = {}
) {
	return callApi<{ status: string }>(
		'update_goal',
		goalId,
		payload.name ?? null,
		payload.target_amount ?? null,
		payload.target_date ?? null,
		payload.status ?? null
	);
}

export function completeGoal(goalId: number) {
	return callApi<{ status: string }>('complete_goal', goalId);
}

export function getGoalAccounts(goalId: number) {
	return callApi<GoalAccount[]>('get_goal_accounts', goalId);
}

export function addAccountToGoal(payload: {
	goal_id: number;
	account_id: number;
	allocated_amount?: number | string;
}) {
	return callApi<{ status: string }>(
		'add_account_to_goal',
		payload.goal_id,
		payload.account_id,
		payload.allocated_amount ?? 0
	);
}

export function updateGoalAccountAllocation(payload: {
	goal_id: number;
	account_id: number;
	allocated_amount: number | string;
}) {
	return callApi<{ status: string }>(
		'update_goal_account_allocation',
		payload.goal_id,
		payload.account_id,
		payload.allocated_amount
	);
}

export function removeAccountFromGoal(goalId: number, accountId: number) {
	return callApi<{ status: string }>('remove_account_from_goal', goalId, accountId);
}
